src/index.js

```diff
@@ -0,0 +1,264 @@
+ * should cancel consuming from any channels, as a new connection will soon
+ * begin consuming.  Errors from such cancellations should be logged and
+ * ignored.  This connection will remain open for 30 seconds to allow any
+ * in-flight message processing to complete.
+ *
+ * The instance will emit `finished` when the connection is finally closed.
+ *
+ * A connection's state can be one of
+ *
+ *  - connecting -- waiting for a connection to complete
+ *  - connected -- connection is up and running
+ *  - retiring -- in the process of retiring
+ *  - finished -- no longer connected
+ *
+ *  Note that an instance that fails to connect will skip from `connecting` to
+ *  `retiring`.
+ *
+ */
+class Connection extends events.EventEmitter {
+  constructor(client, id) {
+    super();
+
+    this.client = client;
+    this.id = id;
+    this.amqp = null;
+
+    this.debug = debug(`taskcluster-lib-pulse:connection:${client.id}.${id}`);
+
+    this.connect();
+  }
+
+  connect() {
+    this.debug('connecting');
+    this.state = 'connecting';
+
+    amqplib.connect(this.client.connectionString, {
+      heartbeat: 120,
+      noDelay: true,
+      timeout: 30 * 1000,
+    }).then(
+      amqp => {
+        if (this.state !== 'connecting') {
+          // we may have been retired already, in which case we do not need this
+          // connection
+          amqp.close();
+          return;
+        }
+        this.amqp = amqp;
+
+        amqp.on('error', err => {
+          if (this.state === 'connected') {
+            this.debug(`error from aqplib connection: ${err}`);
+            this.failed();
+          }
+        });
+
+        amqp.on('close', err => {
+          if (this.state === 'connected') {
+            this.debug('connection closed unexpectedly');
+            this.failed();
+          }
+        });
+
+        this.debug('connected');
+        this.state = 'connected';
+        this.emit('connected');
+      },
+      err => {
+        // TODO: make a minimum interval between connection attempts, to avoid
+        // issue with bad credentials
+        this.debug(`Error while connecting: ${err}`);
+        this.failed();
+      });
+  }
+
+  failed() {
+    if (this.state === 'retired' || this.state === 'finished') {
+      // failure doesn't matter at this point
+      return;
+    }
+    this.debug('failed');
+    this.client.recycle();
+  }
+
+  retire() {
+    if (this.state === 'retiring' || this.state === 'finished') {
+      return;
+    }
+
+    this.debug('retiring');
+    this.state = 'retiring';
+    this.emit('retiring');
+
+    // actually close this connection 30 seconds later
+    setTimeout(() => {
+      this.debug('finished; closing AMQP connection');
+      try {
+        this.amqp.close();
+      } catch (err) {
+        // ignore..
+      }
+      this.amqp = null;
+      this.state = 'finished';
+      this.emit('finished');
+    }, this.client.retirementDelay);
+  }
+}
+
+exports.Connection = Connection;
```

Comment:
Can we avoid exporting this? since this should only be used within this package?

Reply:
It's exported for tests.. if only there was a `pub(crate)` in JS :)